File: Programmers/LEVEL_2/소수찾기.cpp
#include <string>
#include <vector>
#include <cmath>
#include <string>
#include <algorithm>
#include <set>

using namespace std;

bool isPrime(int num) {
    bool flag = true;
    if (num == 0 || num == 1) return false;
    for (int i = 2; i <= sqrt(num); i++)
    {
        if (num % i == 0)
        {
            flag = false;
            break;
        }
    }
    return flag;
}

int solution(string numbers) {
    int answer = 0;
    sort(numbers.begin(), numbers.end(), greater<int>());
    int max_num = stoi(numbers);
    vector<int> prime(max_num + 1);
    vector<bool> visited(max_num + 1);

    for (int i = 0; i <= max_num; i++)
    {
        prime[i] = isPrime(i);
    }

    sort(numbers.begin(), numbers.end());
    set<int> s;
    do {
        for (int i = 1; i <= numbers.size(); i++)
        {
            string temp = numbers.substr(0, i);
            if (prime[stoi(temp)]) {
                s.insert(stoi(temp));
            }
        }

    } while (next_permutation(numbers.begin(), numbers.end()));

    answer = s.size();

    return answer;
}

File: Programmers/LEVEL_2/소수찾기.js
function isPrime(num) {
  if (num < 2) return false;
  if (num == 2) return true;
  for (let i = 2; i <= Math.sqrt(num); i++) {
    if (num % i == 0) return false;
  }
  return true;
}

function solution(numbers) {
  var answer = 0;

  let set = new Set();
  let nums = numbers.split("");

  function permutation(a, s) {
    if (s.length > 0) {
      if (set.has(Number(s)) == false) {
        set.add(Number(s));
        if (isPrime(Number(s))) {
          answer++;
        }
      }
    }

    if (a.length > 0) {
      for (let i = 0; i < a.length; i++) {
        let temp = a.slice(0);
        temp.splice(i, 1);
        permutation(temp, s + a[i]);
      }
    }
  }

  permutation(nums, "");

  return answer;
}
